engine/common/follower/pending_tree/pending_tree.go

```diff
@@ -0,0 +1,198 @@
+//  1. Sorts incoming batch by height. Since blocks can be submitted in random order we need to find blocks with
+//     the lowest height since they are candidates for being connected to the finalized state.
+//  2. Filters out blocks that are already finalized.
+//  3. Deduplicates incoming blocks. We don't store additional vertices in tree if we have that block already stored.
+//  4. Checks for exceeding byzantine threshold. Only one certified block per view is allowed.
+//  5. Finally, blocks with the lowest height from incoming batch that connect to the finalized state we will
+//     mark all descendants as connected, collect them and return as result of invocation.
+//
+// This function is designed to perform resolution of connected blocks(resolved block is the one that connects to the finalized state)
+// using incoming batch. Each block that was connected to the finalized state is reported once.
+// Expected errors during normal operations:
+//   - model.ByzantineThresholdExceededError - detected two certified blocks at the same view
+func (t *PendingTree) AddBlocks(certifiedBlocks []CertifiedBlock) ([]CertifiedBlock, error) {
+	// sort blocks by height, so we can identify if there are candidates for being connected to the finalized state.
+	slices.SortFunc(certifiedBlocks, func(lhs CertifiedBlock, rhs CertifiedBlock) bool {
+		return lhs.Height() < rhs.Height()
+	})
+
+	var allConnectedBlocks []CertifiedBlock
+	for _, block := range certifiedBlocks {
+		// skip blocks lower than finalized view
+		if block.View() <= t.forest.LowestLevel {
+			continue
+		}
+
+		iter := t.forest.GetVerticesAtLevel(block.View())
+		if iter.HasNext() {
+			v := iter.NextVertex().(*PendingBlockVertex)
+
+			if v.VertexID() == block.ID() {
+				// this vertex is already in tree, skip it
+				continue
+			} else {
+				return nil, model.ByzantineThresholdExceededError{Evidence: fmt.Sprintf(
+					"conflicting QCs at view %d: %v and %v",
+					block.View(), v.ID(), block.ID(),
+				)}
+			}
+		}
+
+		vertex, err := NewVertex(block, false)
+		if err != nil {
+			return nil, fmt.Errorf("could not create new vertex: %w", err)
+		}
+		err = t.forest.VerifyVertex(vertex)
+		if err != nil {
+			return nil, fmt.Errorf("failed to store certified block into the tree: %w", err)
+		}
+		t.forest.AddVertex(vertex)
+
+		if t.connectsToFinalizedBlock(block) {
+			connectedBlocks := t.updateAndCollectFork(vertex)
+			allConnectedBlocks = append(allConnectedBlocks, connectedBlocks...)
+		}
+	}
+
+	return allConnectedBlocks, nil
+}
+
+// connectsToFinalizedBlock checks if candidate block connects to the finalized state.
+func (t *PendingTree) connectsToFinalizedBlock(block CertifiedBlock) bool {
+	if block.Block.Header.ParentID == t.lastFinalizedID {
+		return true
+	} else if parentVertex, found := t.forest.GetVertex(block.Block.Header.ParentID); found {
+		return parentVertex.(*PendingBlockVertex).connectedToFinalized
+	} else {
+		return false
+	}
```

Comment:
to me, the following is more readable 

```suggestion
	if block.Block.Header.ParentID == t.lastFinalizedID {
		return true
	}
	if parentVertex, found := t.forest.GetVertex(block.Block.Header.ParentID); found {
		return parentVertex.(*PendingBlockVertex).connectedToFinalized
	}
	return false
```

Reply:
but less efficient, at least we should keep the `else if`.